File: app/src/main/java/com/example/dineout/data/Restaurant.java
package com.example.dineout.data;

import android.os.Parcel;
import android.os.Parcelable;
import com.google.android.gms.maps.model.LatLng;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class Restaurant implements Parcelable {
    private final String id;
    private final String name;
    private final String cuisine;
    private final double rating;
    private final double latitude;
    private final double longitude;
    private final String address;
    private final String phone;
    private final String description;
    private final String priceRange;
    private List<MenuItem> menu;

    public Restaurant(String id, String name, String cuisine, double rating,
                     double latitude, double longitude, String address, String phone,
                     String description, String priceRange) {
        this.id = id;
        this.name = name;
        this.cuisine = cuisine;
        this.rating = rating;
        this.latitude = latitude;
        this.longitude = longitude;
        this.address = address;
        this.phone = phone;
        this.description = description;
        this.priceRange = priceRange;
        this.menu = new ArrayList<>();
    }

    protected Restaurant(Parcel in) {
        id = in.readString();
        name = in.readString();
        cuisine = in.readString();
        rating = in.readDouble();
        latitude = in.readDouble();
        longitude = in.readDouble();
        address = in.readString();
        phone = in.readString();
        description = in.readString();
        priceRange = in.readString();
        menu = new ArrayList<>();
        in.readList(menu, MenuItem.class.getClassLoader());
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(id);
        dest.writeString(name);
        dest.writeString(cuisine);
        dest.writeDouble(rating);
        dest.writeDouble(latitude);
        dest.writeDouble(longitude);
        dest.writeString(address);
        dest.writeString(phone);
        dest.writeString(description);
        dest.writeString(priceRange);
        dest.writeList(menu);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<Restaurant> CREATOR = new Creator<Restaurant>() {
        @Override
        public Restaurant createFromParcel(Parcel in) {
            return new Restaurant(in);
        }

        @Override
        public Restaurant[] newArray(int size) {
            return new Restaurant[size];
        }
    };

    // Getters
    public String getId() { return id; }
    public String getName() { return name; }
    public String getCuisine() { return cuisine; }
    public double getRating() { return rating; }
    public double getLatitude() { return latitude; }
    public double getLongitude() { return longitude; }
    public String getAddress() { return address; }
    public String getPhone() { return phone; }
    public String getDescription() { return description; }
    public String getPriceRange() { return priceRange; }

    public LatLng getLocation() {
        return new LatLng(latitude, longitude);
    }

    public boolean isOpen() {
        // For now, return true as a placeholder
        // In a real app, this would check against opening hours
        return true;
    }

    public static List<Restaurant> getSampleRestaurants() {
        List<Restaurant> restaurants = new ArrayList<>();
        restaurants.add(new Restaurant(
            "1",
            "Greek Taverna",
            "Traditional Greek cuisine",
            4.5,
            37.9715, 23.7267,
            "123 Main St",
            "+30 210 1234567",
            "Authentic Greek dishes in a cozy atmosphere",
            "€€"
        ));
        restaurants.add(new Restaurant(
            "2",
            "Seafood Paradise",
            "Fresh seafood",
            4.7,
            37.9738, 23.7275,
            "456 Ocean Ave",
            "+30 210 7654321",
            "Best seafood in town with sea view",
            "€€€"
        ));
        return restaurants;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Restaurant that = (Restaurant) o;
        return Objects.equals(id, that.id) &&
               Objects.equals(name, that.name) &&
               Objects.equals(cuisine, that.cuisine) &&
               Double.compare(that.rating, rating) == 0 &&
               Objects.equals(priceRange, that.priceRange);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, name, cuisine, rating, priceRange);
    }

    public List<MenuItem> getMenu() {
        return menu;
    }

    public void setMenu(List<MenuItem> menu) {
        this.menu = menu;
    }

    public double calculateDistance(double userLat, double userLng) {
        final int R = 6371; // Earth's radius in kilometers

        double latDistance = Math.toRadians(latitude - userLat);
        double lngDistance = Math.toRadians(longitude - userLng);
        
        double a = Math.sin(latDistance / 2) * Math.sin(latDistance / 2)
                + Math.cos(Math.toRadians(userLat)) * Math.cos(Math.toRadians(latitude))
                * Math.sin(lngDistance / 2) * Math.sin(lngDistance / 2);
        
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        
        return R * c; // Distance in kilometers
    }
} 

File: app/src/main/java/com/example/dineout/data/Restaurant.kt
package com.example.dineout.data

import kotlinx.parcelize.Parcelize
import android.os.Parcelable
import com.google.android.gms.maps.model.LatLng

@Parcelize
data class Restaurant(
    val id: String,
    val name: String,
    val cuisine: String,
    val rating: Double,
    val imageUrl: String? = null,
    val location: LatLng,
    val address: String,
    val phone: String,
    val description: String,
    val priceRange: String,
    val openingHours: Map<String, String>,
    val features: List<String> = emptyList(),
    val menu: List<MenuCategory> = emptyList(),
    val isOpen: Boolean = true,
    val isFavorite: Boolean = false,
    val distance: Double? = null, // in kilometers
    val website: String? = null,
    val email: String? = null,
    val socialMedia: Map<String, String> = emptyMap(),
    val paymentMethods: List<String> = emptyList(),
    val reservations: Boolean = false,
    val delivery: Boolean = false,
    val takeout: Boolean = false,
    val outdoorSeating: Boolean = false,
    val parking: Boolean = false,
    val wifi: Boolean = false,
    val accessibility: Boolean = false,
    val lastUpdated: Long = System.currentTimeMillis()
) : Parcelable {
    fun isOpenNow(): Boolean {
        if (!isOpen) return false
        
        val now = java.util.Calendar.getInstance()
        val currentDay = now.get(java.util.Calendar.DAY_OF_WEEK)
        val currentTime = now.get(java.util.Calendar.HOUR_OF_DAY) * 100 + 
                         now.get(java.util.Calendar.MINUTE)
        
        val dayNames = listOf("Sunday", "Monday", "Tuesday", "Wednesday", 
                            "Thursday", "Friday", "Saturday")
        val currentDayName = dayNames[currentDay - 1]
        
        val hours = openingHours[currentDayName] ?: return false
        val (openTime, closeTime) = hours.split("-").map { it.trim() }
        
        val openHour = openTime.split(":").map { it.toInt() }
        val closeHour = closeTime.split(":").map { it.toInt() }
        
        val openTimeMinutes = openHour[0] * 100 + openHour[1]
        val closeTimeMinutes = closeHour[0] * 100 + closeHour[1]
        
        return currentTime in openTimeMinutes..closeTimeMinutes
    }
    
    fun getAveragePrice(): Double {
        return menu.flatMap { it.items }
            .map { it.price }
            .average()
    }
    
    fun getVegetarianOptions(): List<MenuItem> {
        return menu.flatMap { it.items }
            .filter { it.isVegetarian }
    }
    
    fun getMenuByCategory(categoryId: String): MenuCategory? {
        return menu.find { it.id == categoryId }
    }
}

// Sample restaurants
val sampleRestaurants = listOf(
    Restaurant(
        id = "1",
        name = "Taverna Platanos",
        cuisine = "Greek",
        rating = 4.7,
        address = "15 Dionysiou Areopagitou, Athens 11742",
        phone = "+30 21 0923 8260",
        description = "Traditional Greek taverna serving authentic dishes in a cozy atmosphere with outdoor seating under a huge plane tree.",
        priceRange = "€€",
        location = LatLng(37.9685, 23.7319),
        openingHours = mapOf(
            "Monday" to "12:00 - 23:00",
            "Tuesday" to "12:00 - 23:00", 
            "Wednesday" to "12:00 - 23:00",
            "Thursday" to "12:00 - 23:00",
            "Friday" to "12:00 - 00:00",
            "Saturday" to "12:00 - 00:00",
            "Sunday" to "12:00 - 23:00"
        ),
        features = listOf("Outdoor Seating", "Traditional", "Family Friendly"),
        menu = listOf(
            MenuCategory(
                id = "appetizers1",
                name = "Appetizers",
                description = "Traditional Greek starters to share",
                items = listOf(
                    MenuItem(
                        id = "1",
                        name = "Tzatziki",
                        description = "Creamy yogurt dip with cucumber, garlic and olive oil",
                        price = 4.50,
                        imageUrl = null,
                        isVegetarian = true
                    ),
                    MenuItem(
                        id = "2",
                        name = "Dolmades",
                        description = "Grape leaves stuffed with rice and herbs",
                        price = 5.80,
                        imageUrl = null,
                        isVegetarian = true
                    ),
                    MenuItem(
                        id = "3",
                        name = "Saganaki",
                        description = "Pan-fried cheese with lemon",
                        price = 6.50,
                        imageUrl = null,
                        isVegetarian = true
                    ),
                    MenuItem(
                        id = "4",
                        name = "Greek Salad",
                        description = "Tomatoes, cucumbers, onions, feta cheese and olives",
                        price = 7.50,
                        imageUrl = null,
                        isVegetarian = true
                    )
                )
            ),
            MenuCategory(
                id = "mains1",
                name = "Main Dishes",
                description = "Classic Greek mains cooked with traditional recipes",
                items = listOf(
                    MenuItem(
                        id = "5",
                        name = "Moussaka",
                        description = "Layers of eggplant, potatoes and seasoned ground beef topped with béchamel sauce",
                        price = 12.80,
                        imageUrl = null,
                        isVegetarian = false
                    ),
                    MenuItem(
                        id = "6",
                        name = "Souvlaki Platter",
                        description = "Grilled skewers of marinated pork served with pita, tzatziki and fries",
                        price = 14.50,
                        imageUrl = null,
                        isVegetarian = false
                    ),
                    MenuItem(
                        id = "7",
                        name = "Pastitsio",
                        description = "Baked pasta with ground beef and béchamel sauce",
                        price = 12.00,
                        imageUrl = null,
                        isVegetarian = false
                    ),
                    MenuItem(
                        id = "8",
                        name = "Gemista",
                        description = "Tomatoes and peppers stuffed with rice and herbs",
                        price = 11.50,
                        imageUrl = null,
                        isVegetarian = true
                    )
                )
            ),
            MenuCategory(
                id = "desserts1",
                name = "Desserts",
                description = "Sweet treats to end your meal",
                items = listOf(
                    MenuItem(
                        id = "9",
                        name = "Baklava",
                        description = "Phyllo pastry with nuts and honey",
                        price = 5.50,
                        imageUrl = null,
                        isVegetarian = true
                    ),
                    MenuItem(
                        id = "10",
                        name = "Galaktoboureko",
                        description = "Custard-filled phyllo pastry with syrup",
                        price = 6.00,
                        imageUrl = null,
                        isVegetarian = true
                    ),
                    MenuItem(
                        id = "11",
                        name = "Greek Yogurt with Honey",
                        description = "Creamy yogurt topped with honey and walnuts",
                        price = 4.50,
                        imageUrl = null,
                        isVegetarian = true
                    )
                )
            )
        ),
        isOpen = true,
        isFavorite = false,
        distance = 1.2
    ),
    Restaurant(
        id = "2",
        name = "Psaras Taverna",
        cuisine = "Seafood",
        rating = 4.5,
        address = "22 Aiolou, Athens 10551",
        phone = "+30 21 0321 8733",
        description = "Seafood restaurant specializing in fresh catches of the day with a beautiful view of the harbor.",
        priceRange = "€€",
        location = LatLng(37.9749, 23.7283),
        openingHours = mapOf(
            "Monday" to "18:00 - 23:00",
            "Tuesday" to "18:00 - 23:00", 
            "Wednesday" to "18:00 - 23:00",
            "Thursday" to "18:00 - 23:00",
            "Friday" to "18:00 - 00:00",
            "Saturday" to "12:00 - 00:00",
            "Sunday" to "12:00 - 23:00"
        ),
        features = listOf("Seafood", "Outdoor Seating", "Sea View"),
        menu = listOf(
            MenuCategory(
                id = "appetizers2",
                name = "Appetizers",
                description = "Fresh seafood starters",
                items = listOf(
                    MenuItem(
                        id = "12",
                        name = "Grilled Octopus",
                        description = "Tender octopus grilled with olive oil and lemon",
                        price = 12.00,
                        imageUrl = null,
                        isVegetarian = false
                    ),
                    MenuItem(
                        id = "13",
                        name = "Fried Calamari",
                        description = "Crispy fried calamari served with garlic sauce",
                        price = 8.50,
                        imageUrl = null,
                        isVegetarian = false
                    ),
                    MenuItem(
                        id = "14",
                        name = "Taramasalata",
                        description = "Fish roe dip with olive oil and lemon",
                        price = 5.00,
                        imageUrl = null,
                        isVegetarian = false
                    )
                )
            ),
            MenuCategory(
                id = "mains2",
                name = "Main Dishes",
                description = "Fresh seafood from the Aegean Sea",
                items = listOf(
                    MenuItem(
                        id = "15",
                        name = "Grilled Sea Bream",
                        description = "Fresh sea bream grilled with olive oil, lemon and herbs",
                        price = 18.00,
                        imageUrl = null,
                        isVegetarian = false
                    ),
                    MenuItem(
                        id = "16",
                        name = "Seafood Pasta",
                        description = "Pasta with mixed seafood in tomato sauce",
                        price = 16.50,
                        imageUrl = null,
                        isVegetarian = false
                    ),
                    MenuItem(
                        id = "17",
                        name = "Shrimp Saganaki",
                        description = "Shrimp cooked in tomato sauce with feta cheese",
                        price = 15.00,
                        imageUrl = null,
                        isVegetarian = false
                    )
                )
            )
        ),
        isOpen = true,
        isFavorite = false,
        distance = 0.8
    ),
    Restaurant(
        id = "3",
        name = "To Kafeneio",
        cuisine = "Greek",
        rating = 4.6,
        address = "Loukianou 26, Athens 10675",
        phone = "+30 21 0722 0920",
        description = "A traditional Greek café serving home-style food in a rustic setting with a lovely courtyard.",
        priceRange = "€",
        location = LatLng(37.9783, 23.7414),
        openingHours = mapOf(
            "Monday" to "07:00 - 23:00",
            "Tuesday" to "07:00 - 23:00", 
            "Wednesday" to "07:00 - 23:00",
            "Thursday" to "07:00 - 23:00",
            "Friday" to "07:00 - 00:00",
            "Saturday" to "08:00 - 00:00",
            "Sunday" to "08:00 - 22:00"
        ),
        features = listOf("Breakfast", "Coffee", "Traditional", "Outdoor Seating"),
        menu = listOf(
            MenuCategory(
                id = "breakfast",
                name = "Breakfast",
                description = "Traditional Greek breakfast options",
                items = listOf(
                    MenuItem(
                        id = "18",
                        name = "Greek Yogurt with Honey and Nuts",
                        description = "Thick Greek yogurt with honey and mixed nuts",
                        price = 6.00,
                        imageUrl = null,
                        isVegetarian = true
                    ),
                    MenuItem(
                        id = "19",
                        name = "Spanakopita",
                        description = "Spinach and feta cheese pie in phyllo pastry",
                        price = 4.50,
                        imageUrl = null,
                        isVegetarian = true
                    ),
                    MenuItem(
                        id = "20",
                        name = "Greek Omelet",
                        description = "Eggs with tomatoes, feta cheese and herbs",
                        price = 7.50,
                        imageUrl = null,
                        isVegetarian = true
                    )
                )
            ),
            MenuCategory(
                id = "mains3",
                name = "Main Dishes",
                description = "Home-style Greek favorites",
                items = listOf(
                    MenuItem(
                        id = "21",
                        name = "Beef Stifado",
                        description = "Slow-cooked beef stew with onions and spices",
                        price = 13.50,
                        imageUrl = null,
                        isVegetarian = false
                    ),
                    MenuItem(
                        id = "22",
                        name = "Imam Baildi",
                        description = "Stuffed eggplant with tomato sauce and herbs",
                        price = 10.00,
                        imageUrl = null,
                        isVegetarian = true
                    ),
                    MenuItem(
                        id = "23",
                        name = "Chicken Souvlaki Plate",
                        description = "Grilled chicken skewers with pita, salad and tzatziki",
                        price = 11.50,
                        imageUrl = null,
                        isVegetarian = false
                    )
                )
            ),
            MenuCategory(
                id = "drinks",
                name = "Drinks",
                description = "Beverages to accompany your meal",
                items = listOf(
                    MenuItem(
                        id = "24",
                        name = "Greek Coffee",
                        description = "Traditional Greek coffee served in a small cup",
                        price = 2.50,
                        imageUrl = null,
                        isVegetarian = true
                    ),
                    MenuItem(
                        id = "25",
                        name = "House Wine (500ml)",
                        description = "Local house wine, red or white",
                        price = 8.00,
                        imageUrl = null,
                        isVegetarian = true
                    ),
                    MenuItem(
                        id = "26",
                        name = "Ouzo",
                        description = "Traditional Greek anise-flavored spirit",
                        price = 3.50,
                        imageUrl = null,
                        isVegetarian = true
                    )
                )
            )
        ),
        isOpen = true,
        isFavorite = false,
        distance = 1.5
    )
) 